File: platforms/apple/Sources/Nimbus/Callback.swift
//
// Copyright (c) 2019, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
//

import WebKit

/**
 `Callback` is a native proxy to a javascript function that
 is used for passing callbacks across the bridge.
 */
class Callback: Callable {
    init(webView: WKWebView, callbackId: String) {
        self.webView = webView
        self.callbackId = callbackId
    }

    deinit {
        if let webView = self.webView {
            let script = """
            nimbus.releaseCallback('\(self.callbackId)');
            """
            DispatchQueue.main.async {
                webView.evaluateJavaScript(script)
            }
        }
    }

    func call(args: [Any]) throws -> Any {
        let jsonEncoder = JSONEncoder()
        let jsonArgs = try args.map { arg -> String in
            if let encodable = arg as? Encodable {
                let jsonData = try jsonEncoder.encode(EncodableValue.value(encodable))
                let jsonString = String(data: jsonData, encoding: .utf8)!
                return jsonString
            } else if arg is NSArray || arg is NSDictionary {
                let data = try JSONSerialization.data(withJSONObject: arg, options: [])
                let jsonString = String(data: data, encoding: String.Encoding.utf8)!
                return jsonString
            } else {
                // Parameters passed to callback are implied that they
                // conform to Encodable protocol or be either NSArray or NSDictionary.
                // If for some reason any elements don't throw parameter error.
                throw ParameterError()
            }
        }
        let formattedJsonArgs = String(format: "[%@]", jsonArgs.joined(separator: ","))

        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript("""
                var jsonArgs = \(formattedJsonArgs);
                var mappedJsonArgs = jsonArgs.map(element => {
                  if (element.hasOwnProperty('v')) {
                    return element.v;
                  } else {
                    return element;
                  }
                });
                nimbus.callCallback('\(self.callbackId)', mappedJsonArgs);
            """)
        }
        return ()
    }

    weak var webView: WKWebView?
    let callbackId: String
}

File: platforms/android/nimbus/src/main/java/com/salesforce/nimbus/Callback.kt
//
// Copyright (c) 2019, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
//

package com.salesforce.nimbus

import android.webkit.WebView
import org.json.JSONArray

/**
 * A [Callback] represents a javascript function that can be accepted
 * as an argument to a native method.
 */
class Callback(val webView: WebView, val callbackId: String) {

    protected fun finalize() {
        webView.post {
            webView.evaluateJavascript("__nimbus.releaseCallback('$callbackId');") {}
        }
    }

    /**
     * Execute the wrapped javascript function passing [args] as the arguments.
     */
    fun call(vararg args: Any) {
        val jsonArgs = JSONArray(args)
        webView.post {
            webView.evaluateJavascript("__nimbus.callCallback('$callbackId', $jsonArgs);") {}
        }
    }
}
